mlir/lib/Dialect/Tosa/IR/TosaOps.cpp

```diff
@@ -850,6 +850,71 @@ LogicalResult tosa::ConcatOp::inferReturnTypeComponents(
   return success();
 }
 
+LogicalResult tosa::ConcatOp::verify() {
+  // check that each input has same element type as output
+  auto outType = getOutput().getType();
+  const Operation::operand_range inputList = getInput1();
+
+  if (!llvm::all_of(inputList, [&](auto input) {
+        return succeeded(verifySameElementTypes(
+            *this, /* inType = */ input.getType(), outType));
+      })) {
+    return failure();
+  }
+
+  // Check there is at least one input
+  if (inputList.empty())
+    return emitOpError("expect at least one input");
+
+  const Type firstInputType = inputList.front().getType();
+  const ShapeAdaptor firstInputShape(firstInputType);
+  const int32_t axis = getAxis();
+
+  if (firstInputShape.hasRank()) {
```

Comment:
What if the first input is dynamic but one of the others isn't. Can't we do the axis verification there or we don't want to?

Reply:
Looks like I made the incorrect assumption here that all inputs would have the same rank (or no rank)

Reply:
Sure but what if the first is completely unranked and the rest are ranked. Do we drop verification completely?

Reply:
> Sure but what if the first is completely unranked and the rest are ranked. Do we drop verification completely?

I think so and the `allOperandsHasRank` below is already checking this.

Reply:
I think it's fine to exit early. The spec requires `all inputs must have the same rank`. If the 1st input is totally unranked, it's definitely not verified.

Reply:
I think the case that @GeorgeARM is referring to here is something like:
```
tosa.const(tensor<*xf32>, tensor<1x2x3xf32>)
```
Which is a valid op, until it is determined that the first operand is not rank 3. At this point we should still be able to check that axis makes sense, to try to prove it’s invalid before proceeding further

Reply:
Thanks @lhutton1 for the clarification. Then how about let's move the axis check to the end of the verifier? 

In such case, we'll still do all of the `OperandsHashRank` check. If it hits `tensor<*xf32>`, it will emitError. Once the `OperandsHashRank` check is done. We then check the axis. Then @GeorgeARM 's concern won't happen.

Reply:
In that case we still wouldn’t apply the check for the example above. I think we need to search through all inputs to try to find one with a rank. Once found, we can apply the axis check

Reply:
thanks! updated the code to check this.